### PR TITLE
Refactor code to detect self-referential variables

### PR DESCRIPTION
**Pull request by lizkammer**

### CODE REVIEW

**Review by patricearruda — APPROVED** [review]

File: src/eval.cc [hunk 1]
```diff
@@ -709,19 +709,21 @@ void Evaluator::TraceVariableLookup(const char* operation,
   fprintf(assignment_tracefile_, "    }");
 }
 
-Var* Evaluator::LookupVarForEval(Symbol name) {
-  Var* var = LookupVar(name);
+void Evaluator::AddVarToEvalStack(const Var* var) {
   if (var != nullptr) {
-    if (symbols_for_eval_.find(name) != symbols_for_eval_.end()) {
-      var->SetSelfReferential();
+    if (vars_for_eval_.find(var) != vars_for_eval_.end()) {
```
Comment (lberki) [inline src/eval.cc:714]: When is `var` null?

File: src/var.cc [hunk 2]
```diff
@@ -151,7 +150,9 @@ RecursiveVar::RecursiveVar(Value* v,
 
 void RecursiveVar::Eval(Evaluator* ev, string* s) const {
   ev->CheckStack();
+  ev->AddVarToEvalStack(this);
```
Comment (lberki) [inline src/var.cc:153]: Why not raise the "variable is self-referential" error here? Then one needs to look at fewer places to understand how self-reference detection works. E.g. you could make `AddVarToEvalStack()` return a Boolean or make it signal an error itself.

File: src/var.cc [hunk 3]
```diff
@@ -161,7 +162,7 @@ void RecursiveVar::AppendVar(Evaluator* ev, Value* v) {
 }
 
 void RecursiveVar::Used(Evaluator* ev, const Symbol& sym) const {
-  if (SelfReferential()) {
+  if (ev->SelfReferential(this)) {
```
Comment (lberki) [inline src/var.cc:165]: Remind me again, what causes the recursive evaluation to stop?

**Review by lberki — APPROVED** [review]
Thanks for the change and apologies for the delay.

Please don't consider my comments blockers; this change is an improvement by itself. I'm somewhat embarrassed about not remembering how recursive evaluation is stopped.

File: src/eval.cc (hunk 1 above)
Comment (lizkammer) [reply src/eval.cc:713]: Oops. Forgot to delete this. Thanks.

File: src/var.cc (hunk 2 above)
Comment (lizkammer) [reply src/var.cc:153]: That would be preferable, but unfortunately it's not until the value is marked as used that we get the symbol that we need to use in the error -- I didn't see a straightforward way to pipe through that information without a lot of changes.

File: src/var.cc (hunk 3 above)
Comment (lizkammer) [reply src/var.cc:165]: ERROR_LOC will exit the program with a code of 1.

**Pull request merged**